### PR TITLE
Adding episode v2 api rest endpoint

### PR DESCRIPTION
**Pull request by ratoaq2**
- [x] PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review
- [x] Read [contribution guide](https://github.com/PyMedusa/SickRage/blob/master/.github/CONTRIBUTING.md)
- [x] Added episode v2 api rest endpoint.
- [x] Improved show v2 api rest endpoint.
- [x] Adding log PUT v2 api rest endpoint.

Some examples: (using [httpie](https://github.com/jkbrzt/httpie))

Show information:
- `http http://localhost:8081/api/v2/show/tvdb80379`

```
HTTP/1.1 200 OK
Content-Encoding: gzip
Content-Length: 634
Content-Type: text/html; charset=UTF-8
Date: Sat, 29 Oct 2016 20:39:54 GMT
Etag: "63285065221931fca32647ebd7a2f74cfb375049"
Server: TornadoServer/4.3
Vary: Accept-Encoding

{
    "shows": [
        {
            "airByDate": false, 
            "airs": "Thursday 8:00 PM", 
            "anime": false, 
...
```

Episode information:
- `http  http://localhost:8081/api/v2/show/tvdb80379/s09e01`
- `http  http://localhost:8081/api/v2/show/tvdb80379/2015-09-21/`

```
HTTP/1.1 200 OK
Content-Encoding: gzip
Content-Length: 634
Content-Type: text/html; charset=UTF-8
Date: Sat, 29 Oct 2016 20:39:54 GMT
Etag: "63285065221931fca32647ebd7a2f74cfb375049"
Server: TornadoServer/4.3
Vary: Accept-Encoding

{
    "absoluteNumber": 184, 
    "airDate": "2015-09-21", 
    "description": "Sheldon is confronted by a mystery of the universe he cannot unravel: when a woman wants time apart to think, exactly how much time does that mean, and is there any way to hurry the process along? In Las Vegas, Penny and Leonard march closer to marching down the aisle, but has Penny gotten over Leonard’s infidelity? And if so, will he do anything to un-get her over it?", 
    "episode": 1,
...
```

Episode title
- `http  http://localhost:8081/api/v2/show/tvdb80379/s09e01/title`

```
HTTP/1.1 200 OK
Content-Length: 26
Content-Type: text/html; charset=UTF-8
Date: Sat, 29 Oct 2016 20:40:44 GMT
Etag: "2374119c8d14d7cb75f7c47d056c7fa28c8434f2"
Server: TornadoServer/4.3
Vary: Accept-Encoding

"The Matrimonial Momentum"
```

Episode video metadata:
- `http  http://localhost:8081/api/v2/show/tvdb80379/s09e01/metadata`

```
HTTP/1.1 200 OK
Content-Length: 678
Content-Type: text/html; charset=UTF-8
Date: Sat, 29 Oct 2016 20:41:08 GMT
Etag: "ff4c68cbb2bdcead384846673f8cc8583e3b3b73"
Server: TornadoServer/4.3
Vary: Accept-Encoding

{
    "audio": [
        {
            "bit_rate": 384000, 
            "bit_rate_mode": "Constant", 
            "channels": "5.1", 
            "channels_count": 6, 
            "codec": "AC3", 
...
```

Other examples:
- `http http://localhost:8081/api/v2/log`
- `http http://localhost:8081/api/v2/log/debug`
- `http POST http://localhost:8081/api/v2/log message='Testing logging' level=DEBUG`
- `http POST http://localhost:8081/api/v2/log message='Testing {name}' kwargs:='{"name": ":-)"}'`
- `http POST http://localhost:8081/api/v2/log message='Testing %s' args:='["Wow"]' level=ERROR`
- `http  http://localhost:8081/api/v2/show/tvdb80379/queue`
- `http  http://localhost:8081/api/v2/show/tvdb80379/episodes limit==2 page==2`

@OmgImAlexis

### CODE REVIEW

File: medusa/server/api/v2/base.py [hunk 1]
```diff
@@ -39,4 +44,37 @@ def api_finish(self, status=None, error=None, data=None, headers=None, **kwargs)
...
+        :param fmt:
+        :return:
+        """
+        return BaseRequestHandler._parse(value, lambda d: datetime.strptime(d, fmt))
+
+
+class StringEncoder(json.JSONEncoder):
+    """String json encoder."""
+
+    def default(self, o):
```
Comment (ratoaq2) [inline medusa/server/api/v2/base.py]: Omg codacy... that's ok

**Comment by OmgImAlexis** [comment]
Also glad to see etags being used here. Will that change if the server updates the DB?

**Comment by ratoaq2** [comment]
I didn't touch tvshow apart from the url (because I needed on tvepisode to be consistent). Show endpoint still need to use tvshow.to_json()
I did work on tvepisode using the to_json().

**Comment by ratoaq2** [comment]
What do you think about the urls?

**Comment by ratoaq2** [comment]
The etag is just a hash of the generated json. If a single character changes, the etag will change as well.

File: medusa/server/api/v2/base.py [hunk 2]
```diff
@@ -38,5 +44,60 @@ def api_finish(self, status=None, error=None, data=None, headers=None, **kwargs)
...
```
Comment (fernandog) [inline medusa/server/api/v2/base.py]: @ratoaq2 no idea how to fix this

**Comment by OmgImAlexis** [comment]
The show URLs I'd rather `/show/:id/:episodeId` as each section after the resource explains the next specific area we want. So we want the show resource, specifically the `:id` show and then specifically the `:episodeId` episode.

For the logs I'd like to have `/api/v2/log` return all logs with a page and limit header to paginate through them and a query option to specify the log level.

For example with the headers. 
```
X-Pagination-Count: 100
X-Pagination-Page: 5
X-Pagination-Limit: 20
```

File: medusa/tv.py [hunk 3]
```diff
@@ -1526,62 +1544,116 @@ def __unicode__(self):
...
-            val, key = x.split('::')
-            akas[key] = val
-        return OrderedDict([
-            ('title', self.name),
-            ('akas', akas),
+        indexer_name = reverse_mapping[self.indexer]
+        result = OrderedDict([
             ('id', OrderedDict([
-                ('thetvdb', str(self.indexerid)),  # @TODO: Change this since self.indexerid will be replaced
+                (indexer_name, self.indexerid),
```
Comment (OmgImAlexis) [inline medusa/tv.py]: With this what happens if we want thetvdb and tvmaze for example?

File: medusa/tv.py [hunk 4]
```diff
@@ -1526,62 +1544,116 @@ def __unicode__(self):
...
+            bw_list = self.release_groups or BlackAndWhiteList(self.indexerid)
+            result.update(OrderedDict([
+                ('airs', text_type(self.airs).replace('am', ' AM').replace('pm', ' PM').replace('  ', ' ').strip()),  # e.g Thursday 8:00 PM
+                ('language', self.lang),
+                ('showType', 'sports' if self.is_sports else ('anime' if self.is_anime else 'series')),
+                ('akas', self.get_akas()),
+                ('year', OrderedDict([
+                    ('start', self.imdb_info.get('year') or self.startyear),
+                ])),
+                ('nextAirDate', self.get_next_airdate()),
```
Comment (OmgImAlexis) [inline medusa/tv.py]: Can we get prev and next air dates please.

**Pull request merged**